### ISSUE TITLE
머딥 - 8d

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#37]
# 분류 모델 소개

import warnings
warnings.filterwarnings('ignore')

import pandas as pd
from sklearn.datasets import load_iris
iris = load_iris()

feature_names = iris['feature_names']
feature_names #꽃받침 길이, 너비, 꽃잎 길이, 너비

target = iris['target']
target # 세토사, 베르시컬러, 버지나

df_iris = pd.DataFrame(
    iris['data'],
    columns=iris['feature_names'])

df_iris['target'] = iris['target']

import matplotlib.pyplot as plt
import seaborn as sns

sepal_columns = [df_iris.columns[0], df_iris.columns[1]]

sns.scatterplot(
    x=sepal_columns[1],
    y=sepal_columns[0], 
    hue='target', 
    palette='muted',
    data=df_iris)

petal_columns = [df_iris.columns[2], df_iris.columns[3]]

sns.scatterplot(
    x=petal_columns[1],
    y=petal_columns[0],
    hue='target',
    data=df_iris)

# 데이터 분할
from sklearn.model_selection import train_test_split

X_train, X_valid, y_train, y_valid = train_test_split(
    df_iris.drop('target', axis=1),
    df_iris['target'],
    random_state=42)

X_train.shape, y_train.shape

sns.countplot(y_train)

## 실무 꿀팁!!!!!!!!!!!!
# 회귀분석: 상관계수가 높은 피쳐 기준으로 구간 만들어서 계층 분할
# 분류: 레이블 클래스 비율대로 계층 분류

X_train, X_valid, y_train, y_valid = train_test_split(
    df_iris.drop('target', axis=1),
    df_iris['target'],
    random_state=42,
    stratify=df_iris['target']
    )

sns.countplot(y_train)

# 다중 분류
### k_1 vs k_2, k_1 vs k_3, k_2 vs k_3

from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train, y_train)

pred = model.predict(X_valid)

(pred == y_valid).mean()

model.score(X_valid, y_valid)

## SGD는 한 번에 하나씩 훈련 샘플을 독립적으로 처리한다
from sklearn.linear_model import SGDClassifier

sgd = SGDClassifier(random_state=42)
sgd.fit(X_train, y_train)

pred = model.predict(X_valid)

model.score(X_valid, y_valid)

## 이웃이 많아질 수록 추세를 볼 수 있다.
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()
knn.fit(X_train, y_train)

pred = knn.predict(X_valid)
knn.score(X_valid, y_valid)

## margin을 극대화 추구
from sklearn.svm import SVC

svm = SVC(kernel='linear')
svm.fit(X_train, y_train)

svm.score(X_valid, y_valid)

## 깊이가 높으면 복잡도가 높아진다
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(max_depth=2)
dt.fit(X_train, y_train)

dt.score(X_valid, y_valid)

from sklearn.tree import export_graphviz
from subprocess import call

def graph_tree(model):
    # 파일 내보내기
    export_graphviz(model, out_file='tree.dot')
    # png로 변환
    call(['dot', '-Tpng', 'tree.dot', '-o', 'decistion-tree.png', '-Gdpi=600'])
    # .png 출력
    return Image(filename='decision-tree.png', width=500)

graph_tree(dt)

## confusion matrix
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
import numpy as np

cancer = load_breast_cancer()
data = cancer['data']
target = cancer['target']
feature_names = cancer['feature_names']

df = pd.DataFrame(data=data, columns=feature_names)
df['target'] = target

pos = df.loc[df['target'] == 1]
neg = df.loc[df['target'] == 0]

sample = pd.concat([pos, neg[:5]], sort=True)

X_train, X_valid, y_train, y_valid = train_test_split(
    sample.drop('target', axis=1),
    sample['target'],
    random_state=42
    )

model = LogisticRegression()
model.fit(X_train, y_train)

pred = model.predict(X_valid)
(pred == y_valid).mean() # 97% ?!

bad_prediction = np.ones(shape=y_valid.shape)
(bad_prediction == y_valid).mean() # 98% ?!

from sklearn.metrics import confusion_matrix
import seaborn as sns

sns.heatmap(
    confusion_matrix(y_valid, pred),
    annot=True,
    cmap='Reds'
)
plt.xlabel('predict')
plt.ylabel('actual') # 1=양성, 0=음성(악성)

정밀도 = 양성 중 옳게 분류한 것 (TP / (TP + FP))
재현율(민감도) = 실제 양성 중 옳게 분류한 것 (TP / (TP + FN))
F1 = 참 양성 가중치를 줬을 때 TP 평균

from sklearn.metrics import precision_score
precision_score(y_valid, pred)
precision_score(y_valid, bad_prediction)

from sklearn.metrics import recall_score
recall_score(y_valid, pred)
recall_score(y_valid, bad_prediction)

from sklearn.metrics import f1_score
f1_score(y_valid, pred)
f1_score(y_valid, bad_prediction)
